### PR TITLE
chore: optimize pull-requests workflow

### LINKED ISSUE #756: ci: pull requests workflow should check docs build
The current pull requests workflow doesn't check that the docs build is successful.  
Meaning we would merge a potentially broken build into main, to realize only then it can't be deployed.

### PR DESCRIPTION
**Pull request by sripwoud**
This PR modifies the pull requests GitHub workflow.  
Relevant checks will be run conditionally only if the related src files are changed.

It also includes an additional build checks on `apps/docs` (Closes #756)

## Test Plan
Observe that all the workflow steps unrelated to this PR were skipped.  
Indeed this PR only modifies a yml files in `.github`, observe that it doesn't belong to any glob patterns part of the list of files we want to track.
   So as expected for this PR:
   - test job: all the test job steps are skipped
   image
  - style job: all the steps unrelated to this PR (contracts, subgraph building, eslint...) are also skipped
  image
  
Verify only relevant jobs steps are run when corresponding files are changed (steps aren't skipped when they shouldn't): :heavy_check_mark: see https://github.com/semaphore-protocol/semaphore/pull/765#issuecomment-2097836509

## Related Issue(s)
#756

## Next
Mirror these changes in the production workflow once approach in this PR is agreed on.

## Checklist

-   [x] My code follows the style guidelines of this project
-   [x] I have performed a self-review of my code
-   [x] I have commented my code, particularly in hard-to-understand areas
-   [ ] I have made corresponding changes to the documentation
-   [x] My changes generate no new warnings
-   [x] I have run `yarn format` and `yarn lint` without getting any errors
-   [ ] I have added tests that prove my fix is effective or that my feature works
-   [x] New and existing unit tests pass locally with my changes
-   [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: .lintstagedrc.json [hunk 1]
```diff
@@ -1,4 +1,4 @@
 {
-    "**/*.{js,ts,jsx,tsx,md,json,sol}": "prettier --write",
```
Comment (sripwoud) [inline .lintstagedrc.json:2]: Not directly related to this PR

**Comment by sripwoud** [comment]
@baumstern

File: .github/workflows/pull-requests.yml [hunk 2]
```diff
@@ -22,20 +22,54 @@ jobs:
...
+              run: |
+                  yarn compile:contracts
+                  yarn workspace semaphore-contracts lint
 
-            - name: Compile contracts
-              run: yarn compile:contracts
+            - if: steps.changed-files.outputs.docs_any_changed == 'true'
+              name: Build and format docs
+              run: |
+                  yarn workspace semaphore-docs build
```
Comment (sripwoud) [inline .github/workflows/pull-requests.yml:56]: Closes #756

File: .github/workflows/pull-requests.yml [hunk 3]
```diff
@@ -22,20 +22,54 @@ jobs:
...
+                  files_yaml: |
+                      circuits:
+                        - packages/circuits/**/*.{circom,json,ts}
+                      contracts:
+                        - packages/contracts/**/*.{js,json,ts,sol}
+                      docs:
+                        - apps/docs/**/*
+                      libraries:
+                        - packages/**/*.{js,json,ts}
+                        - '!packages/{circuits/contracts}/**/*'
```
Comment (baumstern) [inline .github/workflows/pull-requests.yml]:
```suggestion
                        - '!packages/{circuits,contracts}/**/*'
```

File: .github/workflows/pull-requests.yml [hunk 4]
```diff
@@ -49,24 +83,45 @@ jobs:
...
+              uses: tj-actions/changed-files@v44
+              with:
+                  files_yaml: |
+                      circuits:
+                        - packages/circuits/**/*.{circom,json,ts}
+                      contracts:
+                        - packages/contracts/**/*.{js,json,ts,sol}
+                      libraries:
+                        - packages/**/*.{js,json,ts}
+                        - '!packages/{circuits/contracts}/**/*'
```
Comment (baumstern) [inline .github/workflows/pull-requests.yml]:
```suggestion
                        - '!packages/{circuits,contracts}/**/*'
```

**Review by baumstern — COMMENTED** [review]
Awesome! Just left some minor comments.

> Testing still left to do: I want to test making some fake changes to the different areas (circuits, contracts etc...) to verify testing/style steps aren't skipped when they shouldn't.

Before I open a PR, I like to make sure everything's checked out by testing different sections—like circuits and contracts—with some dummy changes in my forked repo. Typically, I'll switch the CI trigger from something like `pull_request` trigger to `push` trigger and check if the CI runs succeed for different scenarios. Once everything looks good, I go ahead and open the PR.

You can also run GitHub actions locally using [act](https://github.com/nektos/act). ~~Personally, I prefer to push to a forked repo and observe the remote runs because I am used to pushing several times~~

**Comment by sripwoud** [comment]
> Before I open a PR, I like to make sure everything's checked out by testing different sections—like circuits and contracts—with some dummy changes in my forked repo. Typically, I'll switch the CI trigger from something like pull_request trigger to push trigger and check if the CI runs succeed for different scenarios. Once everything looks good, I go ahead and open the PR.

yes that's more or less what I usually do too. :wink: I also like to open PRs early to get feedbacks early too and avoiding putting to much work into something that may go in the wrong direction.

https://nektosact.com/
this is really nice, I didn't know! Will definitely use it in the future!

**Comment by sripwoud** [comment]
I had a few mistakes in the workflow files that I fixed in the last commits.
These workflows triggered on branches with fake changes demonstrate the expected behavior (what should be skipped is skipped, what should not be skipped is not):
- [trigger-contracts](https://github.com/semaphore-protocol/semaphore/actions/runs/8982386927)
- [trigger-circuits](https://github.com/semaphore-protocol/semaphore/actions/runs/8982586360)
- [trigger-libs](https://github.com/semaphore-protocol/semaphore/actions/runs/8982449994)
- [trigger-subgraph](https://github.com/semaphore-protocol/semaphore/actions/runs/8982848081)
- [trigger-docs](https://github.com/semaphore-protocol/semaphore/actions/runs/8982484158)
- [trigger-lint](https://github.com/semaphore-protocol/semaphore/actions/runs/8982533103)
- [trigger-format](https://github.com/semaphore-protocol/semaphore/actions/runs/8982514223)

**Review by cedoor — APPROVED** [review]
@sripwoud  Amazing <3

Thank you so much for this PR!

I'd say another review should be ok to merge.

**Comment by sripwoud** [comment]
> I'd say another review should be ok to merge.

@baumstern how does it look?

**Comment by sripwoud** [comment]
Let's merge this before #757

**Review by vplasencia — APPROVED** [review]

**Comment by vplasencia** [comment]
Thanks @sripwoud for the great work.

**Review by baumstern — COMMENTED** [review]
LGTM <3

**Pull request merged**